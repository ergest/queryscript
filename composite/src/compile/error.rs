use crate::ast;
use crate::parser::error::ParserError;
use crate::runtime::error::RuntimeError;
use crate::schema::{Decl, Type};
use snafu::{Backtrace, Snafu};
pub type Result<T> = std::result::Result<T, CompileError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum CompileError {
    #[snafu(display("Parser error: {}", source), context(false))]
    SyntaxError {
        source: ParserError,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Parser error: {}", source), context(false))]
    RuntimeError {
        source: RuntimeError,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Unimplemented: {}", what))]
    Unimplemented {
        what: String,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Duplicate entry: {:?}", path))]
    DuplicateEntry {
        path: ast::Path,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("No such entry: {:?}", path))]
    NoSuchEntry {
        path: ast::Path,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display(
        "Wrong kind: declaration at {:?} is {:?} not {}",
        path,
        actual,
        expected
    ))]
    WrongKind {
        path: ast::Path,
        expected: String,
        actual: Decl,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Type mismatch: found {:?} not {:?}", rhs, lhs))]
    WrongType {
        lhs: Type,
        rhs: Type,
        backtrace: Option<Backtrace>,
    },
}

impl CompileError {
    pub fn unimplemented(what: &str) -> CompileError {
        return UnimplementedSnafu { what }.build();
    }

    pub fn no_such_entry(path: ast::Path) -> CompileError {
        return NoSuchEntrySnafu { path }.build();
    }

    pub fn duplicate_entry(path: ast::Path) -> CompileError {
        return DuplicateEntrySnafu { path }.build();
    }

    pub fn wrong_kind(path: ast::Path, expected: &str, actual: &Decl) -> CompileError {
        return WrongKindSnafu {
            path,
            expected,
            actual: actual.clone(),
        }
        .build();
    }

    pub fn wrong_type(lhs: &Type, rhs: &Type) -> CompileError {
        return WrongTypeSnafu {
            lhs: lhs.clone(),
            rhs: rhs.clone(),
        }
        .build();
    }
}
