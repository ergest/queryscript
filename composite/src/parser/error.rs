use snafu::{Backtrace, Snafu};
pub type Result<T> = std::result::Result<T, ParserError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum ParserError {
    #[snafu(display("Unexpected token {:?}: {}", token, msg))]
    UnexpectedToken {
        msg: String,
        token: sqlparser::tokenizer::Token,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("SQL parser error: {}", source), context(false))]
    SQLParserError {
        source: sqlparser::parser::ParserError,
        backtrace: Option<Backtrace>,
    },
}

#[allow(unused_macros)]
macro_rules! unexpected_token {
    ($token: expr, $base: expr $(, $args:expr)* $(,)?) => {
        crate::parser::error::UnexpectedTokenSnafu {
            msg: format!($base $(, $args)*),
            token: $token.clone(),
        }.fail()
    };
}

#[allow(unused_imports)]
pub(crate) use unexpected_token;
