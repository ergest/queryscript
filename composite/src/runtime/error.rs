use snafu::{Backtrace, Snafu};
use std::num::ParseFloatError;
pub type Result<T> = std::result::Result<T, RuntimeError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum RuntimeError {
    #[snafu(display("{}", what))]
    StringError {
        what: String,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Unimplemented: {}", what))]
    Unimplemented {
        what: String,
        backtrace: Option<Backtrace>,
    },

    #[snafu(context(false))]
    ParseFloatError {
        source: ParseFloatError,
        backtrace: Option<Backtrace>,
    },
}

impl RuntimeError {
    pub fn new(what: &str) -> RuntimeError {
        return StringSnafu { what }.build();
    }

    pub fn unimplemented(what: &str) -> RuntimeError {
        return UnimplementedSnafu { what }.build();
    }
}
