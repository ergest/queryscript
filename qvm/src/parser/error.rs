use snafu::{Backtrace, Snafu};
pub type Result<T> = std::result::Result<T, ParserError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum ParserError {
    #[snafu(display("Unexpected token {:?}: {}", token, msg))]
    UnexpectedToken {
        msg: String,
        token: sqlparser::tokenizer::TokenWithLocation,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Tokenizer error: {}", source), context(false))]
    TokenizerError {
        source: sqlparser::tokenizer::TokenizerError,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("SQL parser error: {}", source), context(false))]
    SQLParserError {
        source: sqlparser::parser::ParserError,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Incomplete parse string: {}", original))]
    Incomplete {
        original: Box<ParserError>,
        backtrace: Option<Backtrace>,
    },
}

pub fn wrap_sql_eof<T>(
    result: std::result::Result<T, sqlparser::parser::ParserError>,
) -> Result<T> {
    let wrap = if let Err(sqlparser::parser::ParserError::ParserError(s)) = &result {
        s.ends_with("found: EOF")
    } else if let Err(sqlparser::parser::ParserError::TokenizerError(s)) = &result {
        s.contains("before EOF")
    } else {
        false
    };

    if wrap {
        let err: ParserError = result.err().unwrap().into();
        IncompleteSnafu { original: err }.fail()
    } else {
        Ok(result?)
    }
}

#[allow(unused_macros)]
macro_rules! unexpected_token {
    ($token: expr, $base: expr $(, $args:expr)* $(,)?) => {
        crate::parser::error::UnexpectedTokenSnafu {
            msg: format!($base $(, $args)*),
            token: $token.clone(),
        }.fail()
    };
}

#[allow(unused_imports)]
pub(crate) use unexpected_token;
