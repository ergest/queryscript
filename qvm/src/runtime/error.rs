use snafu::{Backtrace, Snafu};
use std::num::ParseFloatError;
pub type Result<T> = std::result::Result<T, RuntimeError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum RuntimeError {
    #[snafu(display("{}", what))]
    StringError {
        what: String,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("Unimplemented: {}", what))]
    Unimplemented {
        what: String,
        backtrace: Option<Backtrace>,
    },

    #[snafu(context(false))]
    DataFusionError {
        source: datafusion::common::DataFusionError,
        backtrace: Option<Backtrace>,
    },

    #[snafu(context(false))]
    IOError {
        source: std::io::Error,
        backtrace: Option<Backtrace>,
    },

    #[snafu(context(false))]
    ParseFloatError {
        source: ParseFloatError,
        backtrace: Option<Backtrace>,
    },
}

impl RuntimeError {
    pub fn new(what: &str) -> RuntimeError {
        return StringSnafu { what }.build();
    }

    pub fn unimplemented(what: &str) -> RuntimeError {
        return UnimplementedSnafu { what }.build();
    }
}

#[allow(unused_macros)]
macro_rules! fail {
    ($base: expr $(, $args:expr)* $(,)?) => {
        StringSnafu {
            what: format!($base $(, $args)*),
        }.fail()
    };
}

#[allow(unused_imports)]
pub(crate) use fail;

#[allow(unused_macros)]
macro_rules! rt_unimplemented {
    ($base: expr $(, $args:expr)* $(,)?) => {
        UnimplementedSnafu {
            what: format!($base $(, $args)*),
        }.fail()
    };
}

#[allow(unused_imports)]
pub(crate) use rt_unimplemented;
