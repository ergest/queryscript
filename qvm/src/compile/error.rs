use crate::ast;
use crate::compile::schema::{Decl, MType};
use crate::parser::error::ParserError;
use crate::runtime::error::RuntimeError;
use crate::types::error::TypesystemError;
use snafu::{Backtrace, Snafu};
use std::io;
pub type Result<T> = std::result::Result<T, CompileError>;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum CompileError {
    #[snafu(display("Parser error: {}", source), context(false))]
    SyntaxError {
        source: ParserError,
        backtrace: Backtrace,
    },

    #[snafu(display("Internal error: {}", what))]
    InternalError { what: String, backtrace: Backtrace },

    #[snafu(display("Typesystem error: {}", source), context(false))]
    TypesystemError {
        source: TypesystemError,
        backtrace: Backtrace,
    },

    #[snafu(display("Parser error: {}", source), context(false))]
    RuntimeError {
        source: RuntimeError,
        backtrace: Backtrace,
    },

    #[snafu(context(false))]
    FsError {
        source: io::Error,
        backtrace: Backtrace,
    },

    #[snafu(display("Unimplemented: {}", what))]
    Unimplemented { what: String, backtrace: Backtrace },

    #[snafu(display("Duplicate entry: {:?}", path))]
    DuplicateEntry {
        path: ast::Path,
        backtrace: Backtrace,
    },

    #[snafu(display("No such entry: {:?}", path))]
    NoSuchEntry {
        path: ast::Path,
        backtrace: Backtrace,
    },

    #[snafu(display(
        "Wrong kind: declaration at {:?} is {:?} not {}",
        path,
        actual,
        expected
    ))]
    WrongKind {
        path: ast::Path,
        expected: String,
        actual: Decl,
        backtrace: Backtrace,
    },

    #[snafu(display("Type mismatch: found {:?} not {:?}", rhs, lhs))]
    WrongType {
        lhs: MType,
        rhs: MType,
        backtrace: Backtrace,
    },

    #[snafu(display("Error importing {:?}: {}", path, what))]
    ImportError {
        path: ast::Path,
        what: String,
        backtrace: Backtrace,
    },
}

impl CompileError {
    pub fn unimplemented(what: &str) -> CompileError {
        return UnimplementedSnafu { what }.build();
    }

    pub fn no_such_entry(path: ast::Path) -> CompileError {
        return NoSuchEntrySnafu { path }.build();
    }

    pub fn duplicate_entry(path: ast::Path) -> CompileError {
        return DuplicateEntrySnafu { path }.build();
    }

    pub fn wrong_kind(path: ast::Path, expected: &str, actual: &Decl) -> CompileError {
        return WrongKindSnafu {
            path,
            expected,
            actual: actual.clone(),
        }
        .build();
    }

    pub fn wrong_type(lhs: &MType, rhs: &MType) -> CompileError {
        return WrongTypeSnafu {
            lhs: lhs.clone(),
            rhs: rhs.clone(),
        }
        .build();
    }

    pub fn import_error(path: ast::Path, what: &str) -> CompileError {
        return ImportSnafu {
            path,
            what: what.to_string(),
        }
        .build();
    }

    pub fn internal(what: &str) -> CompileError {
        return InternalSnafu {
            what: what.to_string(),
        }
        .build();
    }
}
