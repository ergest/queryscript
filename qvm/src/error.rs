use crate::parser::error::{ErrorLocation, FormattedError, PrettyError};
use snafu::{Backtrace, ErrorCompat, Snafu};
use std::fmt;

#[derive(Debug, Snafu)]
#[snafu(visibility(pub))]
pub enum QVMError {
    #[snafu(display("Syntax error: {}", source), context(false))]
    ParserError {
        #[snafu(backtrace)]
        source: crate::parser::error::ParserError,
    },

    #[snafu(display("Compiler error: {}", source))]
    CompileError {
        #[snafu(backtrace)]
        source: crate::compile::error::CompileError,
    },

    #[snafu(display("Runtime error: {}", source))]
    RuntimeError {
        #[snafu(backtrace)]
        source: crate::runtime::error::RuntimeError,
        file: String,
    },

    #[snafu(whatever, display("{message}"))]
    Whatever {
        message: String,
        #[snafu(source(from(Box<dyn std::error::Error>, Some)))]
        source: Option<Box<dyn std::error::Error>>,
        backtrace: Option<Backtrace>,
    },

    #[snafu(display("{}", sources.first().unwrap()))]
    Multiple {
        // This is assumed to be non-empty
        //
        sources: Vec<QVMError>,
    },
}

impl QVMError {
    pub fn format_backtrace(&self) -> Vec<FormattedError> {
        match self {
            QVMError::Multiple { sources } => {
                let mut ret = Vec::new();
                for source in sources {
                    ret.extend(source.format_backtrace());
                }
                ret
            }
            _ => {
                let mut ret = self.format_without_backtrace()[0].clone();
                if let Some(bt) = ErrorCompat::backtrace(&self) {
                    ret.text += format!("\n{:?}", bt).as_str();
                }
                vec![ret]
            }
        }
    }

    pub fn format_without_backtrace(&self) -> Vec<FormattedError> {
        match self {
            QVMError::Multiple { sources } => {
                let mut ret = Vec::new();
                for source in sources {
                    ret.extend(source.format_without_backtrace());
                }
                ret
            }
            _ => {
                let location = self.location();
                let text = self.to_string();
                vec![FormattedError { location, text }]
            }
        }
    }
}

impl PrettyError for QVMError {
    fn location(&self) -> ErrorLocation {
        match self {
            QVMError::ParserError { source } => source.location(),
            QVMError::CompileError { source } => source.location(),
            QVMError::RuntimeError { file, .. } => ErrorLocation::File(file.clone()),
            _ => ErrorLocation::Unknown,
        }
    }
}

impl<Guard> From<std::sync::PoisonError<Guard>> for QVMError {
    fn from(e: std::sync::PoisonError<Guard>) -> QVMError {
        snafu::FromString::without_source(e.to_string())
    }
}

impl From<crate::compile::error::CompileError> for QVMError {
    fn from(e: crate::compile::error::CompileError) -> QVMError {
        match e {
            crate::compile::error::CompileError::Multiple { sources } => {
                let sources = sources.into_iter().map(|e| e.into()).collect();
                QVMError::Multiple { sources }
            }
            _ => QVMError::CompileError { source: e },
        }
    }
}

pub trait MultiError: fmt::Debug + Sized {
    fn new_multi_error(errs: Vec<Self>) -> Self;
    fn into_errors(self) -> Vec<Self>;
}

#[derive(Debug)]
pub struct MultiResult<V, E: MultiError> {
    pub result: V,
    pub errors: Vec<(Option<usize>, E)>,
}

impl<V, E: MultiError> MultiResult<V, E> {
    pub fn new(result: V) -> MultiResult<V, E> {
        MultiResult {
            result,
            errors: Vec::new(),
        }
    }

    pub fn map<U, F>(self, f: F) -> MultiResult<U, E>
    where
        F: FnOnce(V) -> U,
    {
        MultiResult {
            result: f(self.result),
            errors: self.errors,
        }
    }

    pub fn set_result(&mut self, value: V) {
        self.result = value;
    }

    pub fn add_error(&mut self, idx: Option<usize>, error: E) {
        for err in error.into_errors() {
            self.errors.push((idx, err))
        }
    }

    pub fn absorb<U, E2: MultiError + Into<E>>(&mut self, other: MultiResult<U, E2>) -> U {
        self.errors
            .extend(other.errors.into_iter().map(|(idx, err)| (idx, err.into())));
        other.result
    }

    pub fn replace(&mut self, other: MultiResult<V, E>) {
        self.result = other.result;
        self.errors.extend(other.errors);
    }

    pub fn expect(self, debug: &str) -> V {
        if self.errors.len() == 0 {
            self.result
        } else {
            panic!("{:?}: {:?}", debug, self.errors);
        }
    }

    pub fn unwrap(self) -> V {
        self.expect("error")
    }

    pub fn as_result(mut self) -> Result<V, E> {
        match self.errors.len() {
            0 => Ok(self.result),
            _ => Err(E::new_multi_error(
                self.errors.drain(..).map(|e| e.1).collect(),
            )),
        }
    }

    pub fn ok<'s>(&'s self) -> &'s V {
        &self.result
    }
}

#[macro_export]
macro_rules! c_try {
    ($result: expr, $expr: expr) => {
        match $expr {
            Ok(v) => v,
            Err(e) => {
                $result.add_error(None, e.into());
                return $result;
            }
        }
    };
}
