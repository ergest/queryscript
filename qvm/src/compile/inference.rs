use crate::compile::error::*;
use crate::compile::schema::{mkref, Ref};
use crate::runtime;
use std::fmt;
use std::rc::Rc;

pub trait Constrainable: Clone + fmt::Debug {
    fn unify(&self, other: &Self) -> Result<()>;
}

pub trait Constraint<T: Constrainable>: FnMut(Ref<T>) -> Result<()> {}
pub trait Then<T: Constrainable, R: Constrainable>: FnMut(Ref<T>) -> Result<CRef<R>> {}

impl<T, F> Constraint<T> for F
where
    T: Constrainable,
    F: FnMut(Ref<T>) -> Result<()>,
{
}

impl<T, R, F> Then<T, R> for F
where
    T: Constrainable,
    R: Constrainable,
    F: FnMut(Ref<T>) -> Result<CRef<R>>,
{
}

pub enum Constrained<T>
where
    T: Constrainable,
{
    Known(Ref<T>),
    Unknown {
        debug_name: String,
        constraints: Vec<Box<dyn Constraint<T>>>,
    },
    Ref(CRef<T>),
}

impl<T: Constrainable> fmt::Debug for Constrained<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Constrained::Known(t) => t.borrow().fmt(f),
            Constrained::Unknown { debug_name, .. } => {
                f.write_str(format!("Unknown({:?})", debug_name).as_str())
            }
            Constrained::Ref(r) => r.fmt(f),
        }
    }
}

pub fn mkcref<T: 'static + Constrainable>(t: T) -> CRef<T> {
    CRef::new_known(mkref(t))
}

#[derive(Clone)]
pub struct CRef<T>(Ref<Constrained<T>>)
where
    T: Constrainable;

impl<T: Constrainable> fmt::Debug for CRef<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.borrow().fmt(f)
    }
}

impl<T: 'static + Constrainable> CRef<T> {
    pub fn new_unknown(debug_name: &str) -> CRef<T> {
        CRef(mkref(Constrained::Unknown {
            debug_name: debug_name.to_string(),
            constraints: Vec::new(),
        }))
    }

    pub fn new_known(t: Ref<T>) -> CRef<T> {
        CRef(mkref(Constrained::Known(t)))
    }

    pub fn borrow(&self) -> std::cell::Ref<Constrained<T>> {
        self.0.borrow()
    }

    pub fn borrow_mut(&self) -> std::cell::RefMut<Constrained<T>> {
        self.0.borrow_mut()
    }

    pub fn must(&self) -> runtime::error::Result<Ref<T>> {
        match &*self.find().borrow() {
            Constrained::Known(t) => Ok(t.clone()),
            Constrained::Unknown { .. } => runtime::error::fail!("Unknown cannot exist at runtime"),
            Constrained::Ref(_) => runtime::error::fail!("Canon value should never be a ref"),
        }
    }

    pub fn is_known(&self) -> Result<bool> {
        match &*self.find().borrow() {
            Constrained::Unknown { .. } => Ok(false),
            Constrained::Known(_) => Ok(true),
            _ => Err(CompileError::internal("Canon value should never be a ref")),
        }
    }

    pub fn constrain<F: 'static + Clone + FnMut(Ref<T>) -> Result<()>>(
        &self,
        constraint: F,
    ) -> Result<()> {
        self.add_constraint(Box::new(constraint.clone()))
    }

    pub fn then<R: 'static + Constrainable, F: 'static + Clone + Then<T, R>>(
        &self,
        mut callback: F,
    ) -> Result<CRef<R>> {
        let slot = CRef::<R>::new_unknown("slot");
        let ret = CRef::<R>::new_unknown("then");
        ret.unify(&slot)?;
        let constraint = move |t: Ref<T>| -> Result<()> {
            slot.unify(&callback(t)?)?;
            Ok(())
        };
        self.constrain(constraint)?;

        Ok(ret)
    }

    pub fn unify(&self, other: &CRef<T>) -> Result<()> {
        let us = self.find();
        let them = other.find();

        if Rc::ptr_eq(&us.0, &them.0) {
            return Ok(());
        }

        if !us.is_known()? || !them.is_known()? {
            us.union(&them)?;
        } else {
            us.must()?.borrow().unify(&*them.must()?.borrow())?;
        }

        Ok(())
    }

    // Private methods
    //
    fn add_constraint(&self, mut constraint: Box<dyn Constraint<T>>) -> Result<()> {
        match &mut *self.find().borrow_mut() {
            Constrained::Known(t) => {
                constraint(t.clone())?;
            }
            Constrained::Unknown { constraints, .. } => {
                constraints.push(constraint);
            }
            _ => return Err(CompileError::internal("Canon value should never be a ref")),
        }

        Ok(())
    }

    fn find(&self) -> CRef<T> {
        let new = match &mut *self.borrow_mut() {
            Constrained::Ref(r) => r.find(),
            _ => return self.clone(),
        };

        *self.0.borrow_mut() = Constrained::Ref(new.clone());
        return new;
    }

    fn union(&self, other: &CRef<T>) -> Result<()> {
        if !self.is_known()? && other.is_known()? {
            return other.union(self);
        }

        let us = self.find();
        let them = other.find();

        if !Rc::ptr_eq(&us.0, &them.0) {
            match &mut *them.borrow_mut() {
                Constrained::Unknown { constraints, .. } => {
                    for constraint in constraints.drain(..) {
                        us.add_constraint(constraint)?;
                    }
                }
                _ => {}
            }

            *them.borrow_mut() = Constrained::Ref(us.clone());
        }

        Ok(())
    }
}
