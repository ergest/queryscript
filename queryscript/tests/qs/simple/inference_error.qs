-- This module fails to compile (it references a field that does not exist).
-- It imports the inference module too.
import error;

-- If we re-import inference, we should be able to access the inferred types.
import inference;

SELECT COUNT(*) FROM inference.events;
SELECT COUNT(*) FROM inference.users;
