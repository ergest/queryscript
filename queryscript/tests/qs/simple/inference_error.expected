{
    "compile_errors": [
        (
            None,
            NoSuchEntry {
                path: [
                    "invalid_field",
                ],
                backtrace: None,
            },
        ),
    ],
    "decls": {},
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 4        |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 2        |",
            },
        ),
    ],
}